\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[left=2.54cm, top=3.54cm, right=2.54cm, bottom=2.54cm]{geometry}
\usepackage{amsmath}
\usepackage{graphicx}
\usepackage{subfig}
\usepackage{listings}
\usepackage{hyperref}

%opening
\title{Methods used}
\author{Bram}

\begin{document}

what is CB

why CBC

how CBC

how to measure CBC

separability of targets, before and after CBC

\section{Preprocessing}
%

https://www.ncbi.nlm.nih.gov/pubmed/26036609

between-array correction 
>0 M-values, Type I Green, Type I Red, Type II;
<0 M-values, Type I Green, Type I Red, Type II 

RNA expression: ComBat (L/S plus empirical Bayes): inclusion of covariants using non-parametric empirical Bayes. \\ \\
%
Methylation:
\begin{itemize}
 \item (between-array normalisation): ComBat, parametric empirical Bayes, BEclear
 \item (probewise normalisation): minfi, Infinium I/II correction, color-bias adjustment, background correction
\end{itemize}
%

The TCGA methylation data contains pre-calculated Beta-values, where it is unknown if dye bias and background noise is removed. 
To apply quantile normalisation we should convert the Beta-values to Methylation values. We then apply QN per type, for methylated and unmethylated, after which we re-establish the Beta-values.
%
The so-called M-values are determined as follows:
\begin{equation}
 M  = log2\left({\frac{\beta}{1-\beta}}\right), \quad \beta = \frac{2^M}{2^M+1},
\end{equation}
%
where the methylation is also a direct representation of the amount of methylated versus unmethylated probes according to
\begin{equation}
 M  = log2\left({\frac{max(\gamma_{meth},0)+\alpha}{max(\gamma_{unmeth},0)+\alpha}}\right),
\end{equation}
%
where $\gamma$ is the probe intensity and $\alpha$ (usually set to $1$) is some offset to prevent spurious behavior in case of small $\gamma$. \\ \\
%
The benefit of using M-values as opposed to Beta-values is: 
more homoscedastic than Beta-value distributions, i.e. more suitable for application of techniques
like ANOVA to compare the variances between samples, see e.g. Du et al.\cite{Du2010}.
%
The benefit of the Beta-value is that it has a direct biological interpretation as the \textit{percentage of methylated probes}, however the Beta-values can be directly obtained from the M-values. For this reason Du et al.\cite{Du2010} recommend the use of M-values for differential analysis and Beta-values for biological interpretation. 
%
\section{Cohort bias detection}
%
Before we perform cohort bias removal we seek to quantify the presence of such bias.
We have two general approaches: distribution based and pairwise similarity.

Distribution based: 
\begin{itemize}
 \item Wasserstein metrics
 \item Unsupervised non-parametric statistical significance test: Mann-Whitney U, Kolmogorov-Smirnof
 \item Supervised non-parametric statistical significance test: FDR-ANOVA
\end{itemize}
%
Two options for the application:
1. compare cohorts per feature (or reduced dimension) (columnwise)
2. compare cohorts per patients over the features (or reduced dimensions) (rowwise)

Pairwise similarity:
\begin{itemize}
 \item Kullback-Leibler divergence
 \item Distance metrics/correlation
 \item Bhattacharyya coefficient
\end{itemize}
%
Here, in general we only have on option for the application which is to compare the cohorts per inter-cohort patient-pair.\\\\
When comparing cohorts we have choose to compare each cohort with eachother, or we can compare each cohort with the overall distribution (minus that specific cohort).
%
Classification based:
\begin{itemize}
 \item separation of biological classes by batch identities
\end{itemize}
%
Variation based:
\begin{itemize}
 \item relations between in-group variance, out-group variance and between-group variances,
 see \href{https://www.biorxiv.org/content/biorxiv/early/2014/12/04/012203.full.pdf}{Hicks}
 \item ANOVA and Kruskall-Wallis to check for significantly different distributions.
\end{itemize}

Is between-array normalisation even appropriate given the large variation of biological groups between the cohorts?
According to Dedeurwaerder et al.\cite{Dedeurwaerder2014} when looking differential expressions, the loss in signal is not justified by any benefit that a between-array normalisation can incur. From the same paper it is concluded
that RNA expression correction methods cannot be used as-is to methylation data. The problem being that the different infinium types I and II and the colors within type I are measured on different channels. Also, according to Dedeurwaerder et al. batch effects can generate artifacts that only affect a subset of probes and which thus cannot be corrected with global correction methods. A noted exception is the ComBat\cite{Johnson2007} method as demonstrated in Leek et al.\cite{Leek2010}, Sun et al.\cite{Sun2011} and
Chen et al.\cite{Chen2011}. Nygaard et al.\cite{Nygaard2015}, i.e. a resolution would be to filter out batches that have a significantly different distribution/contribution of relevant biological/pheno types.
\\ \\
%
Within-array normalisation may be justified by the notion that the genetic data may differ over the target variable in terms of the relative importance of type I/II red/green probe values. According to Dedeurwaerder et al.\cite{Dedeurwaerder2014} a Type I/II correction offers the most significant improvement. Hicks and Irizarry\cite{Hicks2014} also question the use of within-array normalisation depending on the nature of the global/between group variation.\\ \\
%
In subsequent papers new methods (other than ComBat) are introduced that do preserve differential expression whilst succesfully removing unwanted bias, methods such as Functional Normalisation, BEclear (detection followed by probe-selective imputance) 
and MANCIE (PCA based).  \\ \\
%
One obvious way to avoid the need for probewise normalisation is too separate the datasets per distinct group:
Type I color red, Type I color green and Type II, standardize per dataset, subsequently apply a cohort-bias removal method such as Limma, ComBat or BEclear. \\ \\
%
We have $36$, $42$  batches for the methylation and RNA expression data respectively. There is a argument to be made for not applying between-array normalisation by virtue of a (roughly) stratification of the covariants over the batches: for the covariants give distributions per batch compared to the overall distribution.

Only apply probewise normalisation to sample groups with little variation (if at all)!
%
So, arguments to \textit{not} use any corrections are 
\begin{itemize}
 \item all types of correction (between/within arrays, PCA-based, linear models, etc.) introduce an unknown bias and may reduce true biological signals 
 \item we can block the batches or particular probes that seem to be affected by batch effects
 \item we can split the methylation data by probe types
 \item for the application of correction methods it is better to use M-values because of the more homoscedastic distributions
 \item given the large number of batches ($36$), even \textit{if} there are significant batch effects, in the aggregate they may not systematically bias the data
\end{itemize}

%
Preliminaries:
\begin{itemize}
 \item what is the common difference in $\beta$-value between genes?
 \item how are the biological groups distributed over the cohorts/samples
 \item how different are the samples from eachother (K-S, MW-U)
 \item what are distances between the distributions and between the batches: Mallowâ€™s distance (1st Wasserstein metric)
 \item what is distribution of the $\Delta \beta$'s over the target groups (cancer-type in this case)
 \item what is per cohort the distribution of median $beta's$ and $\Delta \beta$
 \item how are the PC's clustered per batch/target?
\end{itemize}
%
\section{Cohort bias removal}
%
The following bias removal methods are applied
%
\begin{itemize}
\item RNA expression data: L/S adjustment, cohort based QN, ComBat with the covariate gender
\item Methylation data: ComBat with the covariate gender, BEclear 
\end{itemize}

We apply the cohort bias removal to the measurement cohorts. These cohorts indicate measurement batches and the cohort bias removal reduces any bias that is seemingly related to the cohorts. Arguably we have to apply the bias removal, per cohort, per phenotypical cluster, otherwise the applicability of the cohort bias removal hinges on the degree of stratification of the phenotypes. This is however prohibited by the sparsity of the data.
The ComBat method uses a combination of L/S normalisation/scaling and empirical Bayes to assess the bias that is introduced by the cohort. As a reference we apply L/S, and cohort-wise QN.  \\ 
%
We use the same cohort-bias correction for both the RNA expression data and the methylation data. \\ \\
%
Results are evaluated using:
\begin{itemize}
\item distribution of the log10 of the p-values (K-S, each cohort compared to the bulk), for the FDR we use the current cohort versus the rest as the label
\item distribution of median deviation
\item distribution of mean, max, min 
\item distribution of correlation values between PCA1, PCA2, PCA3
\item plots of (PCA1, PCA2, PCA3), colored by cohort and by target.
\item plots of (UMAP1, UMAP2, UMAP3), colored by cohort and by target.
\item clustering of (sample, sample) similarity (HDBSCAN, AP, MC)
\item differential expression
\item quantile-quantile plots
\end{itemize}
%

The basic observation we should be able to make is the following:
prior to cohort-bias correction the cohort-based clusters should be distinctly seperated, 
and the target based clusters should be distinctly seperated as well. After the CBC the cohort-based clusters
should be significantly more similar. 

For the patient-based clustering we should see an increasing seperation of the different patient groups after the CBC based on the different target values.



\subsection{Batch wise normalisation}
%
Location and scale adjustment (L/S):
\begin{equation}
\mbox{Standard}\quad \mathbf{x}^*_k= \frac{\mathbf{x}_k-\overline{\mathbf{x}}_k}{\sigma_k} + \overline{\mathbf{x}}_k,\quad \forall k\in \mathcal{C}
\end{equation}
%
In literature this approach might be referred to as \textit{standardisation}. \\ \\
%
ComBat, Bayesian based $\rightarrow$ use \href{http://www.bu.edu/jlab/wp-assets/ComBat/Abstract.html}{1ibrary}, part of Bioconductor's sva package. ComBat is a supervised batch effect removal method that requires the explicit input of batches and covariants.\\
%
BEclear, K-S to detect bias-affected batches followed by matrix factorisation techniques to replace suspected batch affected genes in those batches. Downside of BEclear is that it does not take care of the co-variances, upside is that it only applies batch correction to the genes/probes that seem to have a batch effect. When applying this to the methylation data we find approximately 2 million batch affected genes for type I probes and 6 million batch affected genes for type II probes. \\ \\
% 
Functional normalisation, part of Bioconductor's minfi package, function: preprocessFunnorm. An unsupervised 
normalisation method that uses negative control probes. This method is explicitly designed for 450k methylation
data although the idea of negative control probes can be generalised. \\ \\
% 
Alternatively: Concordant bias detection, MANCIE, combining CNV data with expression data.
%https://www.nature.com/articles/ncomms11305
%
To reduce the effect of collinearity we remove all samples that are correlated more than $99\%$ with 
any other sample, we also remove all NaN probe's. \\ 
%
How are the targets distributed over the batches? How do the phenotypical covariants vary within the 
cohorts and between the cohorts? \\ \\
%
To get rid of bias introduced by demographic variations within the cohorts we ideally have a large
independent data set that relates genetic expression data to a wide range of demographic categories, such that research into
demographic dependency of genetic measurement data is structurally open sourced and applied as common bench marks, see e.g. 
Vi\~{n}uela et al\cite{Vinuela2018}.
%
\subsection{Measurement group bias correction}
%
The common approach for probewise (within array) normalisation is some form 
of quantile normalisation. The basic quantile normalisation (QN) 
was succeeded by subset-quantile normalisation (SQN), and smoothed quantile normalisation (SmQN),
, subset-quantile within array normalisation (SWAN) and BMIQ (beta-mixture quantile normalisation).
An alternative for methylation data is the peak-based correction (PBC). \\ \\ 
%
From Wang et al. \cite{Wang2015}: Quantile normalisation replaces the signal intensity of a probe with the mean intensity of the probes that have the same rank from all studied arrays, and thus makes the distribution of probe intensities from each array the same. We are explicitly interested in variance between the target groups, hence we are fine with probewise bias as long as it is roughly stratified over the target groups. 

As an alternative to probewise normalisation it is wiser to simply split the dataset in seperate datasets per probewise group and perform the differential analysis and classification per group: probewise normalisation can reduce the inter-group variation. \\ \\
%
To offset this variance reduction we can apply the probewise normalisation only to the probes that significantly separate the biological groups, say top N\% based on the KS-score. Also, we first transform the beta-values to M-values to increase the homoscedasticity.
We apply Subset Quantile Normalisation by either selecting the top N\% type I probes or the (negative) reference probes;
these reference probes then form the basis for the distribution from which we draw samples for the normalisation. Recall, that
quantile normalisation revolves around the replacement of each rank per sample by the median rank-value of all samples (in the cohort). Instead of using the median rank-values of all samples, with SQN(see Wu and Aryee\cite{Wu2010}) we can estimate a Cumulative Distribution Function by a combination of an empirical (discrete) CDF and $k$ normal distributions as
%
\begin{equation*}
 CDF = (1-w)\hat{CDF} + w\phi
\end{equation*}
%
where $\phi$ is a mixture of normal distributions and $w$ and $k$ are settable parameters to control the smoothness, where Wu and Aryee advice the values $0.9$ and $5$ respectively. 
The current implementation in R is limited to a fixed set of control probes. 

Improved Sheather Jones bandwidth for multimodal KDE with Beta values mirrored around $0$ 
and $1$ (or without mirroring but using M-values). https://github.com/tommyod/KDEpy


\section{Methylation plus RNA expression}
%
We have about $60.000$ RNA expression values, and about $400.000$ methylation values, per sample.
To do a full correlation scan of all combinations we need to perform $60.000\times 400.000 \times 1000$ computations, 
or more specifically, we need to perform $60.000\times 400.000=24\,10^9$ in-products on vectors with length $\propto 1000$. \\ \\
%
To make this tractable we can be selective in the gene's by considering the target variable at hand (say the type 
of cancer) and only select the gene's or probe values that seperate the target variables the best based 
on some non-parametric distribution comparison such as Kolmogorov-Smirnof or Mann-Whitney U, or we apply 
a dimension reduction on both data sets and only directly compare  the top components per datasets.

The caveat with all these approaches is the bias we introduce by considering only the individually strong components per data set, 
The only immediate approach at hand to find the strong combinations of components is a brute-force approach?
Another approach we might try is relatively straightforward: we simply append the components to eachother and apply a
dimension reduction based on the variance (PCA) or the separation (LDA) after we can reconstruct what
components co-occur in the reduced dimensions. \\ \\
%
$(1000, 60.000),(1000, 400.000) \rightarrow (1000, 460.000)$
%
The most biased but biologically sensible approach is the paired combination of methylation and RNA expression data
based on the affected genes. 

\section{Gene/probe distribution comparisons per target}
%
targets are: cancer type, healthy/cancerous tissue, survived/diseased, etc.

\section{Suggestions}

\begin{itemize}
 \item Create ghost batches that are copies of current batches with white noise added
 \item Create ghost batches that are copies of current batches with a bias term added
\end{itemize}
%
\bibliographystyle{plain}
\bibliography{methods}
\end{document}
